### PR TITLE
Step 4 of DGE analysis of matched polya and stranded samples

### PR DESCRIPTION
**Pull request by logstar**
This is the fourth one of the four pull requests for creating the PediatricOpenTargets/OpenPBTA-analysis/analysis/rna-seq-protocol-dge analysis module. These four pull requests correspond to the following four analysis steps:

1. Collapse pbta-gene-counts-rsem-expected_count matrices of stranded and poly-A libraries using the OpenPBTA-analysis/analyses/collapse-rnaseq analysis module.
2. Generate a HTML report of the collapsed genes using the OpenPBTA-analysis/analyses/collapse-rnaseq analysis module.
3. Run edgeR exactTest, glmLRT, and glmQLFTest comparing stranded and poly-A RNA-seq rsem-expected_counts, using an adapted workflow from the OMPARE project.
4. **Select genes that are stably expressed in poly-A and stranded RNA-seq libraries.**

The overall purposes of the PediatricOpenTargets/OpenPBTA-analysis/analysis/rna-seq-protocol-dge module are:

> - Check if the DGE analysis between polya and stranded data follow a null-p-value distribution.
> - Determine stably expressed transcripts between polyA and stranded samples.

#### What scientific question is your analysis addressing?

Determine stably expressed transcripts between polyA and stranded samples.

#### What was your approach?

Select genes that are stably expressed in poly-A and stranded RNA-seq libraries by the following criteria, and the filtered output `results/stranded_vs_polya_stably_exp_genes.csv` is ordered by the coefficient of variations.

   - normalized read count coefficient of variation across all biospecimens < 0.5
   - `exactTest`, `glmLRT`, and `glmQLFTest` p-values all > 0.05
   - logFC < 0.5
   - logCPM > 0.8

#### What GitHub issue does your pull request address?

<https://github.com/PediatricOpenTargets/ticket-tracker/issues/17>

#### Which areas should receive a particularly close look?

Are the criteria for selecting stably expressed genes appropriate? Relevant `03-edger-protocol-dge.R` lines are pasted below.

```R
# Lines 282-290
# select CV < 0.5
seg_df <- seg_df[seg_df$normalized_count_cv < 0.5, ]
# select pval > 0.05
seg_df <- seg_df[seg_df$lrt_pval > 0.05 & seg_df$exact_test_pval > 0.05 &
                     seg_df$qlft_pval > 0.05, ]
# select abs(logFC) < 0.5
seg_df <- seg_df[abs(seg_df$stranded_over_polya_logFC) < 0.5, ]
# select average_logCPM > 0.75
seg_df <- seg_df[seg_df$average_logCPM > 0.8, ]
```

If any revisions should be made to the selection criteria, what additional tables, plots, or statistics could be informative?

#### Is there anything that you want to discuss further?

Are transcripts referring to the rows in the collapsed read count matrices, in the context of this analysis? In <https://github.com/PediatricOpenTargets/ticket-tracker/issues/17>, the analysis purpose is to determine stably expressed transcripts, but the collapsed read count matrices have gene symbols as rows, e.g. RWDD2A, ACAP2, and GGNBP2. The RSEM expected count matrices also have gene symbols and Ensembl gene IDs for each row, e.g. ENSG00000000003.14_TSPAN6, ENSG00000000005.5_TNMD and ENSG00000000419.12_DPM1.

Are we expecting to use read count matrices with ENST transcript IDs for each row?

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes.

#### What types of results are included (e.g., table, figure)?

- Table:
  - `results/stranded_vs_polya_stably_exp_genes.csv`: stably expressed gene table
- Figures:
  - `plots/stably_exp_gene_protocol_diff_boxplot/*png`: normalized read count boxplots
  - `plots/normalized_count_gene_cv_histogram.png`: histogram of coefficient of variations of normalized read counts
  - `plots/edger_logfc_histogram.png`: histogram of edgeR computed log fold changes

#### What is your summary of the results?

The results are summarized in [`README.md`](https://github.com/logstar/OpenPBTA-analysis/tree/rna-seq-protocol-dge-fourth/analyses/rna-seq-protocol-dge).

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.

Note on continuous integration (CI):

- This analysis module currently hardcodes sample IDs, in order to simplify the procedure to find the biospecimens with the same sample_ID and both poly-A and stranded RNA-seq libraries.
- The CI may not be able to run, because the hardcoded sample IDs may not be contained in the CI datasets.
- This CI issue is also discussed in the third pull request of this analysis module (<https://github.com/PediatricOpenTargets/OpenPBTA-analysis/pull/10>).

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

**Comment by logstar** [comment]
Following are the updates on this PR:

**Merge changes in [Step 3 PR](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/10).**

**Change stably expressed gene selection criteria.**

In `03-protocol-dge-seg.R`, select genes that are stably expressed in poly-A and stranded RNA-seq libraries by the following criteria, and the filtered output `results/stranded_vs_polya_stably_exp_genes.csv` is ordered by the coefficient of variations.

- `coefficient of variation < quantile(coefficient of variation, 0.25)`
- `exactTest`, `glmLRT`, `glmQLFTest`, and `nbinomWaldTest` p-values all > 0.05
- `abs(logFC) < median(abs(logFC)`
- `quantile(average_logCPM, 0.25) < average_logCPM < quantile(average_logCPM, 0.75)`

`03-protocol-dge-seg.R` code for selecting stably expressed genes:

```R
seg_df <- seg_df[
    seg_df$normalized_count_cv < quantile(seg_df$normalized_count_cv, 0.25), ]
# select all pval > 0.05
seg_df <- seg_df[seg_df$lrt_pval > 0.05 &
                     seg_df$exact_test_pval > 0.05 &
                     seg_df$qlft_pval > 0.05 &
                     seg_df$deseq2_nbwaldtest_pval > 0.05, ]
seg_df <- seg_df[abs(seg_df$stranded_over_polya_logFC) <
                     median(abs(seg_df$stranded_over_polya_logFC)), ]
# select mean expression between 25th and 75th percentile
alc_idc <- seg_df$average_logCPM > quantile(seg_df$average_logCPM, 0.25)
alc_idc <- alc_idc & (seg_df$average_logCPM < 
                          quantile(seg_df$average_logCPM, 0.75))
seg_df <- seg_df[alc_idc, ]
# order by CV
seg_df <- seg_df[order(seg_df$normalized_count_cv), ]
rownames(seg_df) <- NULL
```

**Comment by logstar** [comment]
Merged changes in [Step 3 PR](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/10).

**Review by aadamk — APPROVED** [review]
Really interesting results here @logstar. 
My take: 
1. UQ-pgQ2 works well in this setting, since there is a balanced experimental design in this case. This is evidenced by the results for the QL F Test, where there are the fewest number of significant hits passing the FDR threshold relative to all other normalization + statistical test combinations. 
2. I agree with the updated criteria for selecting stably expressed genes (intersection across tests + cv method, etc). 
3. The gene list in `uqpgq2_normalized/stranded_vs_polya_stably_exp_genes.csv` will likely be suitable for downstream batch correction modeling via RUVseq for future DGE comparisons. Validating this would require benchmarking of a dataset for which there are no expected differences to test for a uniform distribution of p-values (when correcting for batch using this gene list).

Thanks so much!

**Comment by logstar** [comment]
> Really interesting results here @logstar.
> My take:
> 
> 1. UQ-pgQ2 works well in this setting, since there is a balanced experimental design in this case. This is evidenced by the results for the QL F Test, where there are the fewest number of significant hits passing the FDR threshold relative to all other normalization + statistical test combinations.
> 2. I agree with the updated criteria for selecting stably expressed genes (intersection across tests + cv method, etc).
> 3. The gene list in `uqpgq2_normalized/stranded_vs_polya_stably_exp_genes.csv` will likely be suitable for downstream batch correction modeling via RUVseq for future DGE comparisons. Validating this would require benchmarking of a dataset for which there are no expected differences to test for a uniform distribution of p-values (when correcting for batch using this gene list).
> 
> Thanks so much!

Thank you for the detailed review!

File: analyses/rna-seq-protocol-dge/README.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    9     | BS_7WM3MNZ0               | A16915    | RNA-Seq               | poly-A      | PNOC003
+    10    | BS_KABQQA0T               | A16915    | RNA-Seq               | stranded    | PNOC003
+    11    | BS_68KX6A42               | A18777    | RNA-Seq               | poly-A      | PNOC003
+    12    | BS_D7XRFE0R               | A18777    | RNA-Seq               | stranded    | PNOC003
+3. Run edgeR `exactTest`, edgeR `glmLRT`, edgeR `glmQLFTest`, and DESeq2 `nbinomWaldTest` comparing stranded and poly-A RNA-seq `rsem-expected_count`s, using an adapted workflow from the [OMPARE project](https://github.com/d3b-center/OMPARE/blob/master/code/patient_level_analyses/utils/rnaseq_edger_normalizations.R#L37-L41). RSEM expected read counts are normalized using either UQ-pgQ2 or TMM. The UQ-pgQ2 normalization method has lower false positive rate in DGE testing (<https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-020-6502-7>).
+4. Select genes that are stably expressed in poly-A and stranded RNA-seq libraries by the following criteria, and the filtered output `results/stranded_vs_polya_stably_exp_genes.csv` is ordered by the coefficient of variations.
```
Comment (kgaonkar6) [inline analyses/rna-seq-protocol-dge/README.md]: should it be results/NORMALIZATION_METHOD/stranded_vs_polya_stably_exp_genes.csv ?

```suggestion
4. Select genes that are stably expressed in poly-A and stranded RNA-seq libraries by the following criteria, and the filtered output `results/NORMALIZATION_METHOD/stranded_vs_polya_stably_exp_genes.csv` is ordered by the coefficient of variations. 
```

**Review by kgaonkar6 — APPROVED** [review]
Looks great! Once the conflicts are resolved ok to merge.

File: analyses/rna-seq-protocol-dge/README.md (hunk 1 above)
Comment (logstar) [reply analyses/rna-seq-protocol-dge/README.md]: Yes. Thank you for catching it.

**Comment by logstar** [comment]
> Looks great! Once the conflicts are resolved ok to merge.

Thank you for the review!

**Pull request merged**